### PR TITLE
[query service] example read_table works

### PR DESCRIPTION
**Pull request by cseed**
@danking @tpoterba I wasn't sure if this should go to compilers or services team.  I randomly picked Dan.  Who "owns" GoogleStorageFS?  ServiceBackend?  Let's discuss in staff meeting this week.

Summary of changes:
 - ExecuteContext aded to IRParserEnvironment (maybe should be renamed IRParserContext now?)
 - TableReader in general will need the filesystem for construct.  This breaks the json4s de/serializer model.  I started expanding the serializers and writing them by hand.  Just did the TableNativeReader for now.  We might reconsider the mix of custom and JSON in the printed format at some point in the future.
 - A bunch of functions take a hc but only need an fs.  More of this to come.

```
hl.init(_backend=hl.backend.ServiceBackend())
t = hl.read_table('gs://hail-cseed-k0ox4/sample_rows.ht')
print(t.count())
```

```
Welcome to
     __  __     <>__
    / /_/ /__  __/ /
   / __  / _ `/ / /
  /_/ /_/\_,_/_/_/   version 0.2.36-75a0f869d72d
LOGGING: writing to /home/cotton/hail-20200407-1502-0.2.36-75a0f869d72d.log
346
```

### CODE REVIEW

File: hail/src/main/scala/is/hail/expr/ir/NativeReaderOptions.scala [hunk 1]
```diff
@@ -31,8 +28,31 @@ class NativeReaderOptionsSerializer() extends CustomSerializer[NativeReaderOptio
     })
 )
 
+object NativeReaderOptions {
+  def fromJson(optionsJV: JObject): NativeReaderOptions = {
```
Comment (danking) [inline hail/src/main/scala/is/hail/expr/ir/NativeReaderOptions.scala:32]: I really like this naming convention. I feel like I'm always hunting for how to (de)serialize an object.

File: hail/src/main/scala/is/hail/expr/ir/Parser.scala [hunk 2]
```diff
@@ -138,6 +138,7 @@ case class TypeParserEnvironment(
 }
 
 case class IRParserEnvironment(
+  ctx: ExecuteContext,
```
Comment (danking) [inline hail/src/main/scala/is/hail/expr/ir/Parser.scala:141]: IIUC, you only need the `FS` from this, right? Seems weird to put something called "ExecuteContext" into the parser environment. This is Tim's wheelhouse though, so I don't fully understand the intentions of IRParserEnvironment.

File: hail/src/test/scala/is/hail/expr/ir/IRSuite.scala [hunk 3]
```diff
@@ -2887,6 +2888,7 @@ class IRSuite extends HailSuite {
   @Test(dataProvider = "matrixIRs")
   def testMatrixIRParser(x: MatrixIR) {
     val s = Pretty(x, elideLiterals = false)
+    println(s)
```
Comment (danking) [inline hail/src/test/scala/is/hail/expr/ir/IRSuite.scala:2891]: Rogue println. Should this be the message of the assert below?

**Comment by danking** [comment]
Totally unrelated: what ends up in `/home/cotton/hail-20200407-1502-0.2.36-75a0f869d72d.log`? What happens to the usual Spark master logs?

File: hail/src/main/scala/is/hail/expr/ir/Parser.scala (hunk 2 above)
Comment (cseed) [reply hail/src/main/scala/is/hail/expr/ir/Parser.scala:141]: Yeah, it is weird.  Putting an FS in it also seems weird, but that's the reality of having an IR node with a path whose type is determined by contents of the file at that path.  I can pull out just the FS.  The reference genomes will also move to the execute context and is needed by the parser, so this seems appropriate, but happy to break them out individual if you prefer.

File: hail/src/main/scala/is/hail/expr/ir/NativeReaderOptions.scala (hunk 1 above)
Comment (cseed) [reply hail/src/main/scala/is/hail/expr/ir/NativeReaderOptions.scala:32]: Thanks!  Yeah, I agree, the json serialization in particular seems uber-implicit.

**Comment by cseed** [comment]
> what ends up in /home/cotton/hail-20200407-1502-0.2.36-75a0f869d72d.log? What happens to the usual Spark/Hail master logs?

The client and the server are not separated by a machine boundary, so just the Python client logs end up in the client log.  That's basically nothing and can probably be removed when using the client with the service.

There are no Spark logs, the service is 100% Spark-free.  The Hail master logs end up in the query service logs.

Obviously a lot of this needs to be rethough and improved.  The error checking and reporting needs to get improved at the service boundary, errors should be relative to the input, and clients probably shouldn't get a server-side stack trace.  We're going to need additional tools for debugging pipelines on the master, and probably want an admin UI where you get the logs for each query, the IR getting executed, how it was transformed with lowering/optimization, statistics on timing of its execution, etc.

**Pull request merged**